### ISSUE TITLE
[MEDIUM] ML-Based Opportunity Scoring

### ISSUE THREAD

**Issue by ilan-pinto** [ilan-pinto/alc-lite#7]
## Overview
Implement machine learning-based opportunity scoring to enhance the existing Global Opportunity Selection system. Use historical performance data to improve arbitrage opportunity selection accuracy.

## Problem Statement
Current rule-based scoring may miss optimal patterns:
- Static weights don't adapt to market conditions
- Missing complex feature interactions
- No learning from historical success/failure patterns
- Limited market regime awareness

## Acceptance Criteria
- [ ] 20% improvement in win rate over current scoring
- [ ] Real-time inference under 50ms per opportunity
- [ ] A/B testing framework with statistical significance
- [ ] Model retraining pipeline for market adaptation

### Feature Engineering Pipeline
```python
from dataclasses import dataclass
from typing import List, Dict, Optional
import pandas as pd
import numpy as np

@dataclass
class OpportunityFeatures:
    # Market Microstructure Features
    bid_ask_spread_ratio: float
    volume_imbalance: float
    quote_intensity: float
    market_depth: float
    
    # Greeks and Sensitivities
    net_delta: float
    net_gamma: float
    net_theta: float
    net_vega: float
    pin_risk_score: float
    
    # Market Regime Features
    vix_level: float
    vix_percentile: float
    term_structure_slope: float
    sector_iv_rank: float
    correlation_to_market: float
    
    # Strategy-Specific Features
    profit_potential: float
    max_loss_potential: float
    time_to_expiry: float
    days_to_earnings: Optional[int]
    
    # Historical Context
    symbol_success_rate: float
    strategy_recent_performance: float
    market_regime_performance: float

class FeatureEngineer:
    def __init__(self):
        self.vix_lookback = 252  # 1 year for percentile calculation
        self.correlation_window = 20  # 20-day correlation
        
    async def extract_features(self, opportunity: Dict, market_data: Dict) -> OpportunityFeatures:
        """Extract comprehensive features for ML model"""
        
        # Market microstructure
        bid_ask_spread = (opportunity['ask'] - opportunity['bid']) / opportunity['mid']
        volume_ratio = opportunity['bid_size'] / (opportunity['bid_size'] + opportunity['ask_size'])
        
        # Greeks calculation
        greeks = await self.calculate_opportunity_greeks(opportunity)
        
        # Market regime detection
        vix_current = market_data['vix']
        vix_history = market_data['vix_history']
        vix_percentile = self.calculate_percentile(vix_current, vix_history)
        
        # Historical performance lookup
        symbol_history = await self.get_symbol_performance_history(opportunity['symbol'])
        
        return OpportunityFeatures(
            bid_ask_spread_ratio=bid_ask_spread,
            volume_imbalance=volume_ratio,
            quote_intensity=market_data.get('quote_updates_per_minute', 0),
            market_depth=min(opportunity['bid_size'], opportunity['ask_size']),
            
            net_delta=greeks.delta,
            net_gamma=greeks.gamma,
            net_theta=greeks.theta,
            net_vega=greeks.vega,
            pin_risk_score=self.calculate_pin_risk(opportunity),
            
            vix_level=vix_current,
            vix_percentile=vix_percentile,
            term_structure_slope=market_data.get('term_structure_slope', 0),
            sector_iv_rank=market_data.get('sector_iv_rank', 0.5),
            correlation_to_market=market_data.get('market_correlation', 0),
            
            profit_potential=opportunity['max_profit'],
            max_loss_potential=opportunity['max_loss'],
            time_to_expiry=opportunity['days_to_expiry'],
            days_to_earnings=market_data.get('days_to_earnings'),
            
            symbol_success_rate=symbol_history.get('win_rate', 0.5),
            strategy_recent_performance=symbol_history.get('recent_performance', 0.5),
            market_regime_performance=symbol_history.get('regime_performance', 0.5)
        )
```

### Model Architecture
```python
import xgboost as xgb
from sklearn.ensemble import RandomForestRegressor
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense, Dropout
import joblib

class EnsembleOpportunityScorer:
    def __init__(self):
        self.xgb_model = None
        self.lstm_model = None
        self.ensemble_weights = {'xgb': 0.6, 'lstm': 0.4}
        self.feature_scaler = StandardScaler()
        
    def build_models(self):
        """Build ensemble of XGBoost and LSTM models"""
        
        # XGBoost for feature importance and fast inference
        self.xgb_model = xgb.XGBRegressor(
            n_estimators=300,
            max_depth=6,
            learning_rate=0.1,
            subsample=0.8,
            colsample_bytree=0.8,
            random_state=42,
            objective='reg:squarederror'
        )
        
        # LSTM for temporal pattern recognition
        self.lstm_model = Sequential([
            LSTM(64, return_sequences=True, input_shape=(30, 25)),  # 30 days, 25 features
            Dropout(0.2),
            LSTM(32, return_sequences=False),
            Dropout(0.2),
            Dense(16, activation='relu'),
            Dense(1, activation='sigmoid')  # Output probability score
        ])
        
        self.lstm_model.compile(
            optimizer='adam',
            loss='mse',
            metrics=['mae']
        )
    
    async def train_models(self, training_data: pd.DataFrame, labels: pd.Series):
        """Train both models on historical data"""
        
        # Prepare features
        feature_columns = [col for col in training_data.columns if col not in ['outcome', 'timestamp']]
        X = training_data[feature_columns]
        y = labels
        
        # Train XGBoost
        X_scaled = self.feature_scaler.fit_transform(X)
        self.xgb_model.fit(X_scaled, y)
        
        # Prepare temporal data for LSTM
        X_temporal = self.prepare_temporal_features(training_data)
        
        # Train LSTM
        self.lstm_model.fit(
            X_temporal, y,
            epochs=50,
            batch_size=32,
            validation_split=0.2,
            verbose=1
        )
        
        # Calculate feature importance
        self.feature_importance = dict(zip(
            feature_columns,
            self.xgb_model.feature_importances_
        ))
    
    async def predict_score(self, features: OpportunityFeatures) -> float:
        """Generate ensemble prediction for opportunity score"""
        
        # Convert to array format
        feature_array = self.features_to_array(features)
        feature_array_scaled = self.feature_scaler.transform([feature_array])
        
        # XGBoost prediction
        xgb_score = self.xgb_model.predict(feature_array_scaled)[0]
        
        # LSTM prediction (requires temporal context)
        temporal_features = await self.get_temporal_context(features)
        lstm_score = self.lstm_model.predict(temporal_features)[0][0]
        
        # Ensemble prediction
        final_score = (
            self.ensemble_weights['xgb'] * xgb_score +
            self.ensemble_weights['lstm'] * lstm_score
        )
        
        return np.clip(final_score, 0.0, 1.0)
```

### Online Learning System
```python
class OnlineLearningManager:
    def __init__(self, model_scorer: EnsembleOpportunityScorer):
        self.scorer = model_scorer
        self.feedback_buffer = []
        self.retrain_threshold = 100  # Retrain after 100 new samples
        
    async def record_outcome(self, opportunity_id: str, features: OpportunityFeatures, 
                           actual_outcome: float):
        """Record actual trading outcome for model learning"""
        
        self.feedback_buffer.append({
            'opportunity_id': opportunity_id,
            'features': features,
            'outcome': actual_outcome,
            'timestamp': datetime.now()
        })
        
        # Trigger retraining if threshold reached
        if len(self.feedback_buffer) >= self.retrain_threshold:
            await self.incremental_retrain()
    
    async def incremental_retrain(self):
        """Incrementally update models with new data"""
        
        # Prepare new training data
        new_data = pd.DataFrame(self.feedback_buffer)
        
        # Update XGBoost with new samples
        # (XGBoost supports incremental learning via warm_start)
        X_new = self.prepare_features(new_data)
        y_new = new_data['outcome']
        
        # Retrain with both old and new data (weighted)
        await self.scorer.incremental_fit(X_new, y_new)
        
        # Clear buffer
        self.feedback_buffer = []
        
        # Log retraining event
        logger.info(f"Model retrained with {len(new_data)} new samples")
```

### A/B Testing Framework
```python
class ABTestingFramework:
    def __init__(self):
        self.control_scorer = RuleBasedScorer()  # Current system
        self.treatment_scorer = MLBasedScorer()   # New ML system
        self.test_split = 0.5  # 50/50 split
        self.results_buffer = []
        
    async def score_opportunity(self, opportunity: Dict) -> Tuple[float, str]:
        """Route opportunity to control or treatment group"""
        
        # Deterministic assignment based on opportunity hash
        assignment_hash = hash(f"{opportunity['symbol']}_{opportunity['timestamp']}")
        use_treatment = (assignment_hash % 100) < (self.test_split * 100)
        
        if use_treatment:
            score = await self.treatment_scorer.score(opportunity)
            group = "treatment"
        else:
            score = await self.control_scorer.score(opportunity)
            group = "control"
        
        # Record assignment for analysis
        self.results_buffer.append({
            'opportunity_id': opportunity['id'],
            'group': group,
            'score': score,
            'timestamp': datetime.now(),
            'features': opportunity
        })
        
        return score, group
    
    def analyze_results(self) -> Dict[str, float]:
        """Analyze A/B test results for statistical significance"""
        
        df = pd.DataFrame(self.results_buffer)
        
        control_outcomes = df[df['group'] == 'control']['actual_outcome']
        treatment_outcomes = df[df['group'] == 'treatment']['actual_outcome']
        
        # Statistical significance test
        from scipy.stats import ttest_ind
        t_stat, p_value = ttest_ind(treatment_outcomes, control_outcomes)
        
        return {
            'control_mean': control_outcomes.mean(),
            'treatment_mean': treatment_outcomes.mean(),
            'improvement': (treatment_outcomes.mean() - control_outcomes.mean()) / control_outcomes.mean(),
            'p_value': p_value,
            'statistically_significant': p_value < 0.05,
            'sample_size': len(df)
        }
```

### Model Monitoring & Drift Detection
```python
class ModelMonitor:
    def __init__(self):
        self.baseline_distribution = None
        self.drift_threshold = 0.1  # PSI threshold
        
    def calculate_psi(self, expected_dist: np.ndarray, actual_dist: np.ndarray) -> float:
        """Calculate Population Stability Index (PSI) for drift detection"""
        
        # Avoid division by zero
        expected_dist = np.clip(expected_dist, 1e-10, 1)
        actual_dist = np.clip(actual_dist, 1e-10, 1)
        
        psi = np.sum((actual_dist - expected_dist) * np.log(actual_dist / expected_dist))
        return psi
    
    async def check_model_drift(self, recent_features: List[OpportunityFeatures]) -> Dict[str, float]:
        """Monitor for feature drift that indicates model degradation"""
        
        if self.baseline_distribution is None:
            return {'drift_detected': False, 'psi_score': 0.0}
        
        # Convert features to distribution
        recent_dist = self.features_to_distribution(recent_features)
        
        # Calculate PSI
        psi_score = self.calculate_psi(self.baseline_distribution, recent_dist)
        
        drift_detected = psi_score > self.drift_threshold
        
        if drift_detected:
            logger.warning(f"Model drift detected! PSI score: {psi_score}")
            # Trigger model retraining or alert
            await self.trigger_retraining_alert()
        
        return {
            'drift_detected': drift_detected,
            'psi_score': psi_score,
            'threshold': self.drift_threshold
        }
```

### Enhanced Global Selection
```python
class MLEnhancedGlobalSelection:
    def __init__(self, ml_scorer: EnsembleOpportunityScorer, 
                 ab_tester: ABTestingFramework):
        self.ml_scorer = ml_scorer
        self.ab_tester = ab_tester
        self.fallback_scorer = RuleBasedScorer()  # Fallback for failures
        
    async def select_best_opportunity(self, opportunities: List[Dict]) -> Optional[Dict]:
        """Enhanced opportunity selection with ML scoring"""
        
        scored_opportunities = []
        
        for opp in opportunities:
            try:
                # Get ML-enhanced score through A/B testing
                score, group = await self.ab_tester.score_opportunity(opp)
                
                scored_opportunities.append({
                    **opp,
                    'ml_score': score,
                    'test_group': group,
                    'scoring_method': 'ml_enhanced'
                })
                
            except Exception as e:
                # Fallback to rule-based scoring
                logger.warning(f"ML scoring failed for {opp['symbol']}: {e}")
                fallback_score = await self.fallback_scorer.score(opp)
                
                scored_opportunities.append({
                    **opp,
                    'ml_score': fallback_score,
                    'test_group': 'fallback',
                    'scoring_method': 'rule_based'
                })
        
        # Select highest scoring opportunity
        if not scored_opportunities:
            return None
            
        best_opportunity = max(scored_opportunities, key=lambda x: x['ml_score'])
        
        return best_opportunity
```

### Real-time Inference
```python
class FastInferenceEngine:
    def __init__(self, model_path: str):
        # Load optimized model (ONNX for faster inference)
        import onnxruntime as ort
        self.ort_session = ort.InferenceSession(model_path)
        
    async def predict_batch(self, features_batch: List[OpportunityFeatures]) -> List[float]:
        """Batch prediction for improved throughput"""
        
        # Convert to numpy array
        feature_arrays = [self.features_to_array(f) for f in features_batch]
        input_data = np.array(feature_arrays, dtype=np.float32)
        
        # ONNX inference
        ort_inputs = {self.ort_session.get_inputs()[0].name: input_data}
        predictions = self.ort_session.run(None, ort_inputs)[0]
        
        return predictions.tolist()
```

### Model Versioning
```python
class ModelVersionManager:
    def __init__(self, model_registry_path: str):
        self.registry_path = model_registry_path
        self.current_version = None
        
    async def deploy_new_version(self, model, version: str, validation_results: Dict):
        """Deploy new model version with validation"""
        
        # Validate model performance
        if validation_results['accuracy'] < 0.8:
            raise ValueError("Model accuracy too low for deployment")
        
        # Save model with version
        model_path = f"{self.registry_path}/model_v{version}"
        joblib.dump(model, f"{model_path}.pkl")
        
        # Update metadata
        metadata = {
            'version': version,
            'deployed_at': datetime.now().isoformat(),
            'validation_results': validation_results,
            'model_path': model_path
        }
        
        with open(f"{model_path}_metadata.json", 'w') as f:
            json.dump(metadata, f)
        
        self.current_version = version
        logger.info(f"Deployed model version {version}")
```

### Model Validation
```python
def test_model_accuracy():
    """Validate model accuracy on held-out test set"""
    
    # Load test data
    test_data = load_test_dataset()
    X_test = test_data.drop(['outcome'], axis=1)
    y_test = test_data['outcome']
    
    # Model predictions
    scorer = EnsembleOpportunityScorer()
    predictions = scorer.predict_batch(X_test)
    
    # Calculate metrics
    from sklearn.metrics import accuracy_score, precision_score, recall_score
    
    accuracy = accuracy_score(y_test > 0.5, predictions > 0.5)
    precision = precision_score(y_test > 0.5, predictions > 0.5)
    recall = recall_score(y_test > 0.5, predictions > 0.5)
    
    assert accuracy > 0.8, f"Model accuracy {accuracy} below threshold"
    assert precision > 0.75, f"Model precision {precision} below threshold"
    assert recall > 0.70, f"Model recall {recall} below threshold"

def test_inference_speed():
    """Ensure inference meets latency requirements"""
    
    scorer = FastInferenceEngine("model.onnx")
    
    # Create sample features
    sample_features = [create_sample_features() for _ in range(100)]
    
    start_time = time.time()
    predictions = scorer.predict_batch(sample_features)
    elapsed = time.time() - start_time
    
    avg_latency = elapsed / len(sample_features) * 1000  # ms per prediction
    
    assert avg_latency < 50, f"Average latency {avg_latency}ms exceeds 50ms limit"
```

## Success Metrics
- 20% improvement in opportunity selection win rate
- <50ms inference latency for real-time scoring
- 90% model uptime with graceful fallback
- Statistical significance in A/B tests within 30 days

## Dependencies
- Requires backtesting framework (Issue #2) for training data
- Integration with existing Global Opportunity Selection
- Performance monitoring for model drift detection

## Priority: P1 (Medium)
Significant value add but not blocking other development.

**Estimated Effort**: 4-5 weeks
**Assignee**: TBD
**Milestone**: Phase 3 - Q3 2025
